### DISCUSSION TITLE
Jest says "ReferenceError: Request is not defined" at node_modules/next/src/server/web/spec-extension/request.ts

### DISCUSSION THREAD

**Discussion by moonman239** [vercel/next.js#59041] [Help] [answered]
### Summary

I'm trying to test an API endpoint by testing its GET function, but when I run Jest, my test fails because Request is apparently not defined.

Any help would be greatly appreciated!

### Additional information

```js
File under test:

import { NextRequest, NextResponse } from "next/server";

export async function GET(request: NextRequest): Promise<NextResponse>
{
    return new NextResponse("");
}
```

Test file:

```
import { NextRequest } from "next/server";
import { GET } from "./route";
import { Message } from "./types";
test("GET /",async ()=>{
    process.env.SECRET_KEY = "";
    process.env.BASE_URL = "https://mockgpt.wiremockapi.cloud/v1";
    const messages: Message[] = [{role:"user",content:"Hi!"}];
    const request = new NextRequest("/",{
        body: JSON.stringify(messages)
    });
```

Test failure:
```
 FAIL  app/chatbot/route.test.ts
  ● Test suite failed to run

    ReferenceError: Request is not defined

      at Object.<anonymous> (node_modules/next/src/server/web/spec-extension/request.ts:10:34)
      at Object.<anonymous> (node_modules/next/server.js:2:16)
```
```


### Example

_No response_

**Comment by icyJoseph** [comment, marked as answer]
- https://github.com/mswjs/msw/issues/686#issuecomment-844539072

So this is in part, trying to understand, how Jest tests work.

I assume you have `jsdom` setup in your `jest.config.js` file? 

Try adding this to your test:

```js
/**
 * @jest-environment node
 */
```

So that your test looks like:

```js
/**
 * @jest-environment node
 */

import { NextRequest } from "next/server";
import { GET } from "./route";
import { Message } from "./types";
test("GET /",async ()=>{
    process.env.SECRET_KEY = "";
    process.env.BASE_URL = "https://mockgpt.wiremockapi.cloud/v1";
    const messages: Message[] = [{role:"user",content:"Hi!"}];
    const request = new NextRequest("/",{
        body: JSON.stringify(messages)
    });
```

The default is `node`, but in your `jest.config.js`, you are probably changing it to `jsdom`(jest-environment-jsdom), and that means that you need to opt-out of that, when you want the test to run in Node.js...

Try with that. There's also projects, https://jestjs.io/docs/configuration#projects-arraystring--projectconfig, which can help you to tell Jest, how to run certain tests files.

**Reply by moonman239** [reply]
I don't have a jest.config.js, I have a jest.config.mjs:

```
import nextJest from 'next/jest.js'
 
const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: './',
})
 
// Add any custom config to be passed to Jest
/** @type {import('jest').Config} */
const config = {
  // Add more setup options before each test is run
  // setupFilesAfterEnv: ['<rootDir>/jest.setup.js'],
 
  testEnvironment: 'jest-environment-jsdom',
}
 
// createJestConfig is exported this way to ensure that next/jest can load the Next.js config which is async
export default createJestConfig(config)
```

**Reply by eric-burel** [reply]
For the record "projects" doesn't seem to work properly with `next/js`. I used to have a [killer setup for fullstack testing](https://github.com/VulcanJS/vulcan-npm/blob/main/jest.config.js#L224) but it doesn't work anymore. This [thread from 2023](https://github.com/vercel/next.js/discussions/31152#discussioncomment-4049567) seems to mention the issue but with no solution.
So switching the environment on the fly using a comment does the trick for now.

**Comment by EcksDy** [comment]
For anyone coming from google:

If your test needs to be in `jsdom` environment, there's an [issue with `jest-environment-jsdom`](https://github.com/jsdom/jsdom/issues/1724), where they didnt implement fetch API.
You can use [whatwg-fetch](https://www.npmjs.com/package/whatwg-fetch) polyfill lib, to make sure the fetch API is available in the test environment.

Install the lib as dev dependency:
`npm i -D whatwg-fetch`

Add the following to `jest.setup.js`:
```
setupFiles: ['<rootDir>/setup-jest.js'],
```

Add the following to the head of `setup-jest.js`
```
import 'whatwg-fetch';
```
